californium-core/src/main/java/org/eclipse/californium/core/network/BaseMatcher.java

```diff
@@ -184,7 +187,6 @@ protected final Exchange matchNotifyResponse(final ExchangeObserver exchangeObse
 			request.setDestinationPort(response.getSourcePort());
 			exchange = new Exchange(request, Origin.LOCAL, obs.getContext());
 			exchange.setRequest(request);
-			exchange.setObserver(exchangeObserver);
```

Comment:
The blockwise exchange only needs cleanup, if further blocks are read. But in that case, the implementations of Matcher.sendRequest() will set the ExchangeObserver at the point, where the Exchange is added to the matchers lists.